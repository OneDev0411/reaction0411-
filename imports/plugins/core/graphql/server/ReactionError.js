import ExtendableError from "es6-error";

export default class ReactionError extends ExtendableError {
  constructor(error, message, eventData = {}) {
    super(message);
    this.eventData = eventData;
    this.error = error;
  }
}
