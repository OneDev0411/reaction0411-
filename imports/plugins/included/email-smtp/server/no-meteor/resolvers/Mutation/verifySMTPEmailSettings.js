/**
 * @name Mutation/verifySMTPEmailSettings
 * @method
 * @summary resolver for the verifySMTPEmailSettings GraphQL mutation
 * @param {Object} _ - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} [input.host] - The host of the SMTP email service. Automatically provided if service !== `custom`.
 * @param {String} input.password - password of SMTP email service account.
 * @param {String} [input.port] - The port of the SMTP email service. Automatically provided if service !== `custom`.
 * @param {String} input.service - SMTP email service to use.
 * @param {String} input.shopId - ShopID this setting belongs to.
 * @param {String} input.user - username of SMTP email service account.
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @returns {Promise<Object>} verifySMTPEmailSettingsPayload
 */
export default async function verifySMTPEmailSettings(_, { input }, context) {
  const {
    clientMutationId = null,
    host,
    password,
    port,
    service,
    shopId,
    user
  } = input;

  const verified = await context.mutations.verifySMTPEmailSettings(context, {
    host,
    password,
    port,
    service,
    shopId,
    user
  });

  return {
    clientMutationId,
    verified
  };
}
