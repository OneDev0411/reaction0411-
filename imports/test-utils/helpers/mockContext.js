const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getAbsoluteUrl: jest.fn().mockName("getAbsoluteUrl").mockImplementation((path) => {
    const adjustedPath = path[0] === "/" ? path : `/${path}`;
    return `https://app.mock${adjustedPath}`;
  }),
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  mutations: {},
  queries: {},
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

/**
 * @summary Returns a mock collection instance with the given name
 * @param {String} collectionName The collection name
 * @returns {Object} Mock collection instance
 */
export function mockCollection(collectionName) {
  return {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    bulkWrite: jest.fn().mockName(`${collectionName}.bulkWrite`).mockReturnValue(Promise.resolve({
      nMatched: 2,
      nModified: 2,
      result: {
        writeErrors: []
      }
    })),
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    findOneAndDelete: jest.fn().mockName(`${collectionName}.findOneAndDelete`),
    findOneAndUpdate: jest.fn().mockName(`${collectionName}.findOneAndUpdate`),
    fetch: jest.fn().mockName(`${collectionName}.fetch`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
}

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "ExampleIOUPaymentRefunds",
  "Groups",
  "MediaRecords",
  "NavigationItems",
  "NavigationTrees",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "SimpleInventory",
  "Tags",
  "Templates",
  "Themes",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = mockCollection(collectionName);
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
