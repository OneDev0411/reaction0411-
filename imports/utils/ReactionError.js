import ExtendableError from "es6-error";

/**
 * @name ReactionError
 * @class
 * @memberof GraphQL
 * @summary A type of error that allows additional details to be provided, which can
 *   then be sent back to the client for a GraphQL request.
 */
export default class ReactionError extends ExtendableError {
  constructor(error, message, eventData = {}) {
    super(message);
    this.eventData = eventData;
    this.error = error;
    // Newer versions of DDP use this property to signify that an error
    // can be sent back and reconstructed on the calling client.
    this.isClientSafe = true;
    // DDP expects this to be in a property named `reason`
    this.reason = message;
    // DDP expects this to be in a property named `details`
    this.details = eventData;
  }

  clone() {
    return new ReactionError(this.error, this.reason, this.eventData);
  }
}
