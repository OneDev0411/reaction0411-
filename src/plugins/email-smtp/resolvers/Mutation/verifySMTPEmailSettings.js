/**
 * @name Mutation/verifySMTPEmailSettings
 * @method
 * @summary resolver for the verifySMTPEmailSettings GraphQL mutation
 * @param {Object} _ - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} input.shopId - ShopID this setting belongs to.
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @returns {Promise<Object>} verifySMTPEmailSettingsPayload
 */
export default async function verifySMTPEmailSettings(_, { input }, context) {
  const {
    clientMutationId = null,
    shopId
  } = input;

  const isVerified = await context.mutations.verifySMTPEmailSettings(context, {
    shopId
  });

  return {
    clientMutationId,
    isVerified
  };
}
