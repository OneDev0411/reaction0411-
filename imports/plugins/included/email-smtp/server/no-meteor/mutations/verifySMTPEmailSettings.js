import SimpleSchema from "simpl-schema";
import Logger from "@reactioncommerce/logger";
import nodemailer from "@reactioncommerce/nodemailer";
import getServiceConfig from "nodemailer-wellknown";
import ReactionError from "@reactioncommerce/reaction-error";
import getMailConfig from "../util/getMailConfig";

const inputSchema = new SimpleSchema({
  host: {
    type: String,
    optional: true
  },
  password: String,
  port: {
    type: String,
    optional: true
  },
  service: String,
  shopId: String,
  user: String
});

/**
 * @name verifySMTPEmailSettings
 * @summary Verify the current email configuration
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - an object of all mutation arguments that were sent by the client
 * @param {String} [input.host] - The host of the SMTP email service. Automatically provided if service !== `custom`.
 * @param {String} input.password - password of SMTP email service account.
 * @param {String} [input.port] - The port of the SMTP email service. Automatically provided if service !== `custom`.
 * @param {String} input.service - SMTP email service to use.
 * @param {String} input.shopId - ShopID this setting belongs to.
 * @param {String} input.user - username of SMTP email service account.
 * @returns {Boolean} - returns true if SMTP connection succeeds
 */
export default async function verifySMTPEmailSettings(context, input) {
  inputSchema.validate(input);

  const { userHasPermission } = context;
  const { host, port, service, shopId, user, password } = input;

  if (!userHasPermission(["owner", "admin", "dashboard"], shopId)) throw new ReactionError("access-denied", "Access denied");

  let config;

  // if a settings object has been provided, build a config
  if (typeof input === "object") {
    if (service === "custom" && host && port) {
      // create a custom Nodemailer config
      config = { host, port };

      if (host === "localhost") {
        config.ignoreTLS = true;
      }
    } else if (service) {
      // create a Nodemailer config from the nodemailer-wellknown services
      config = getServiceConfig(service) || {};
    }

    if (user && password) {
      config.auth = { user, pass: password };
    }
  }

  if (!config) {
    config = await getMailConfig(context, shopId);
  }

  const logConfig = { ...config };
  delete logConfig.auth;
  Logger.debug(logConfig, "Verifying email config settings");

  const transporter = nodemailer.createTransport(config);

  const verified = await transporter.verify()
    .then((result) => result)
    .catch((error) => { throw new ReactionError(error.responseCode, error.response); });

  return verified;
}
