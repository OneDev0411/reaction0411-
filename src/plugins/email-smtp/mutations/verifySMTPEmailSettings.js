import SimpleSchema from "simpl-schema";
import Logger from "@reactioncommerce/logger";
import nodemailer from "@reactioncommerce/nodemailer";
import ReactionError from "@reactioncommerce/reaction-error";
import getMailConfig from "../util/getMailConfig.js";

const inputSchema = new SimpleSchema({
  shopId: String
});

/**
 * @name verifySMTPEmailSettings
 * @summary Verify the current email configuration
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - an object of all mutation arguments that were sent by the client
 * @param {String} input.shopId - ShopID this setting belongs to.
 * @returns {Boolean} - returns true if SMTP connection succeeds
 */
export default async function verifySMTPEmailSettings(context, input) {
  inputSchema.validate(input);

  const { checkPermissions } = context;
  const { shopId } = input;

  await checkPermissions(["owner", "admin", "dashboard"], shopId);

  const config = await getMailConfig();

  const logConfig = { ...config };
  delete logConfig.auth;
  Logger.debug(logConfig, "Verifying email config settings");

  const transporter = nodemailer.createTransport(config);

  let isVerified;
  try {
    isVerified = await transporter.verify();
  } catch (error) {
    throw new ReactionError(error.responseCode, error.response);
  }

  return isVerified;
}
