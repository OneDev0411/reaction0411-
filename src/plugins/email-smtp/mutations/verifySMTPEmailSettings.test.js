import mockContext from "@reactioncommerce/api-utils/tests/mockContext.js";
import ReactionError from "@reactioncommerce/reaction-error";
import verifySMTPEmailSettings from "./verifySMTPEmailSettings.js";

beforeEach(() => {
  jest.resetAllMocks();
});

const shopId = "SHOP_ID";

test("throws if permission check fails", async () => {
  mockContext.checkPermissions.mockImplementation(() => {
    throw new ReactionError("access-denied", "Access Denied");
  });

  await expect(verifySMTPEmailSettings(mockContext, {
    shopId
  })).rejects.toThrowErrorMatchingSnapshot();

  expect(mockContext.checkPermissions).toHaveBeenCalledWith(["owner", "admin", "dashboard"], "SHOP_ID");
});
