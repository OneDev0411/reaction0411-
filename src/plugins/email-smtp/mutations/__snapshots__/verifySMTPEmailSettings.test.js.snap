// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if permission check fails 1`] = `"Access Denied"`;
